File: prototype/operations.js
(() => {
  "use strict";
  const $ = (id) => document.getElementById(id);
  const labels = {ok: "조회 성공", no_index: "인덱스 없음", unavailable: "조회 실패", not_started: "실행 기록 없음", stale: "상태 갱신 지연 / 중단 여부 확인 필요", success: "최근 정규화 성공", running: "정규화 실행 중", failed: "정규화 실패", waiting: "다음 수신 구간 대기"};
  const evidenceLabels = {exact_reference: "정확한 문서 참조 확인 (내용 해시 재검증은 아님)", normalized_missing_or_expired: "정규화 근거 없음 / 만료", raw_missing_or_expired: "원본 근거 없음 / 만료", provenance_mismatch: "근거 참조 불일치", unsupported_or_invalid_reference: "지원하지 않거나 잘못된 근거 참조", invalid_reference: "잘못된 근거 참조"};
  const fmt = (value) => value ? new Date(value).toLocaleString("ko-KR", {timeZone: "Asia/Seoul", hour12: false}) : "미확인";
  let generation = 0, detailGeneration = 0;
  function node(tag, value, className) {
    const element = document.createElement(tag);
    if (value !== undefined) element.textContent = value;
    if (className) element.className = className;
    return element;
  }
  async function request(path) {
    const response = await fetch(path, {cache: "no-store", credentials: "same-origin"});
    if (!response.ok) throw new Error("서버 연결·권한·조회 조건을 확인하세요. (" + response.status + ")");
    return response.json();
  }
  function chart(target, section) {
    target.replaceChildren();
    if (!section || section.state !== "ok") { target.append(node("p", labels[section?.state] || "미확인")); return; }
    if (!section.count) { target.append(node("p", "조회 구간의 기록이 없습니다.")); return; }
    const buckets = section.buckets, max = Math.max(1, ...buckets.map((item) => item.count));
    const svg = document.createElementNS("http://www.w3.org/2000/svg", "svg");
    svg.setAttribute("viewBox", "0 0 600 140"); svg.setAttribute("role", "img");
    svg.setAttribute("aria-label", "기간별 문서 수. 총 " + section.count + "건");
    buckets.forEach((item, i) => {
      const rect = document.createElementNS("http://www.w3.org/2000/svg", "rect");
      const height = item.count / max * 130;
      for (const [key, value] of Object.entries({x: i * 600 / buckets.length, y: 140 - height, width: Math.max(0.5, 600 / buckets.length - 1), height})) rect.setAttribute(key, String(value));
      const title = document.createElementNS("http://www.w3.org/2000/svg", "title");
      title.textContent = fmt(item.time) + " · " + item.count + "건"; rect.append(title); svg.append(rect);
    });
    const box = node("div", undefined, "ops-bars"); box.append(svg); target.append(box, node("p", fmt(buckets[0]?.time) + " ~ " + fmt(buckets.at(-1)?.time)));
  }
  function render(data) {
    if (!data || !data.sections || !Number.isFinite(Date.parse(data.start)) || !Number.isFinite(Date.parse(data.end))) throw new Error("관제 응답 형식을 확인하세요.");
    const sections = data.sections;
    for (const key of ["intake", "processing", "alerts", "quality"]) {
      const s = sections[key];
      if (!s || !["ok", "no_index", "unavailable"].includes(s.state) || (s.state === "ok" &&
          (!Number.isSafeInteger(s.count) || s.count < 0 || !Array.isArray(s.rows) || s.rows.length > 50 || !Array.isArray(s.statuses) ||
           !Array.isArray(s.buckets) || s.buckets.length > 1000 || s.buckets.some(b => !Number.isFinite(Date.parse(b.time)) || !Number.isSafeInteger(b.count) || b.count < 0)))) throw new Error("관제 응답 형식을 확인하세요.");
    }
    if (!sections.pipeline || !Object.hasOwn(labels, sections.pipeline.state)) throw new Error("처리기 응답 형식을 확인하세요.");
    $("ops-range").textContent = fmt(data.start) + " ~ " + fmt(data.end) + " (KST)";
    const metrics = $("ops-metrics"); metrics.replaceChildren();
    for (const [key, title] of [["intake", "수신 로그"], ["processing", "처리 결과 기록"], ["alerts", "저장된 경보"], ["quality", "수집 품질 보고"]]) {
      const section = sections[key], card = node("article", undefined, "panel");
      card.append(node("h2", title), node("strong", section.state === "ok" ? section.count.toLocaleString("ko-KR") : "미확인", "ops-value"), node("p", labels[section.state] || "미확인"));
      if (key === "processing" && section.state === "ok") card.append(node("p", section.statuses.map((s) => ({normalized: "정규화", unsupported: "미지원", invalid: "검증 실패"}[s.key] || "기타") + " " + s.doc_count).join(" · ") || "처리 기록 없음"));
      metrics.append(card);
    }
    const pipeline = sections.pipeline;
    $("ops-pipeline").textContent = "정규화 처리기: " + (labels[pipeline.state] || "미확인") + " · 마지막 성공: " + fmt(pipeline.last_success) + " · 수신 시각 체크포인트: " + fmt(pipeline.checkpoint);
    chart($("ops-intake-chart"), sections.intake); chart($("ops-alert-chart"), sections.alerts);
    const alerts = sections.alerts; $("ops-alert-rows").replaceChildren();
    $("ops-alert-state").textContent = alerts.state === "ok" ? "총 " + alerts.count + "건 · 최신 최대 50건 표시 (페이지 탐색 미지원)" : labels[alerts.state];
    for (const row of alerts.rows || []) {
      const tr = node("tr");
      const severity = {critical: "긴급", high: "높음", medium: "중간", low: "낮음"}[row.severity] || row.severity || "미확인";
      for (const value of [fmt(row.timestamp), severity, row.rule, row.title, row.organization]) tr.append(node("td", value || "미관측"));
      const td = node("td"), button = node("button", "상세 / 근거");
      button.type = "button"; button.addEventListener("click", () => detail(row.id)); td.append(button); tr.append(td); $("ops-alert-rows").append(tr);
    }
  }
  async function detail(id) {
    const version = ++detailGeneration;
    $("ops-detail-content").replaceChildren(); $("ops-detail-state").textContent = "조회 중";
    if (!$("ops-detail").open) $("ops-detail").showModal();
    try {
      const data = await request("/api/alerts/detail?" + new URLSearchParams({id}));
      if (version !== detailGeneration) return;
      $("ops-detail-state").textContent = data.evidence_count ? "저장된 근거 " + data.evidence_count + "건 · 최대 100건 개별 확인" : "과거 경보에 정확한 근거 참조가 없습니다. 주변 로그로 대체하지 않습니다.";
      const condition = data.condition;
      $("ops-detail-content").append(node("p", (data.alert.title || data.alert.rule || "경보") + " · 규칙 버전 " + (data.rule_version || "미기록")), node("p", "관측 수 " + (condition.event_count ?? "미기록") + " / 임계값 " + (condition.threshold ?? "미기록") + " / 시간 창 " + (condition.time_window_seconds ?? "미기록") + "초"));
      for (let i = 0; i < data.evidence_limit; i++) {
        const box = node("section", undefined, "ops-evidence"), button = node("button", "근거 " + (i + 1) + " 확인"), result = node("div");
        button.type = "button";
        button.addEventListener("click", async () => {
          button.disabled = true; result.textContent = "조회 중";
          try {
            const reply = await request("/api/alerts/detail?" + new URLSearchParams({id, evidence: String(i)}));
            if (version !== detailGeneration) return;
            const item = reply.evidence;
            result.replaceChildren(node("p", evidenceLabels[item.state] || "미확인"));
            if (item.raw) result.append(node("p", "원본: " + item.raw.index + " / " + item.raw.id));
            if (item.metadata) result.append(node("pre", JSON.stringify(item.metadata, null, 2)));
          } catch (error) { result.textContent = error.message; } finally { button.disabled = false; }
        });
        box.append(button, result); $("ops-detail-content").append(box);
      }
    } catch (error) { if (version === detailGeneration) $("ops-detail-state").textContent = error.message; }
  }
  async function refresh() {
    const version = ++generation;
    $("ops-error").hidden = true; $("ops-range").textContent = "조회 중";
    // Clear stale counts while loading a different interval or after failure.
    for (const id of ["ops-metrics", "ops-intake-chart", "ops-alert-chart", "ops-alert-rows"]) $(id).replaceChildren();
    $("ops-pipeline").textContent = "처리기 상태 조회 중"; $("ops-alert-state").textContent = "조회 중";
    const end = new Date(), start = new Date(end.getTime() - Number($("ops-period").value) * 60000);
    try {
      const data = await request("/api/operations?" + new URLSearchParams({start: start.toISOString(), end: end.toISOString()}));
      if (version === generation) render(data);
    } catch (error) {
      if (version !== generation) return;
      for (const id of ["ops-metrics", "ops-intake-chart", "ops-alert-chart", "ops-alert-rows"]) $(id).replaceChildren();
      $("ops-error").hidden = false; $("ops-error").textContent = error.message;
      $("ops-range").textContent = "조회 실패"; $("ops-pipeline").textContent = "처리기 상태 미확인"; $("ops-alert-state").textContent = "조회 실패";
    }
  }
  $("ops-filter").addEventListener("submit", (event) => {event.preventDefault(); refresh();});
  $("ops-detail-close").addEventListener("click", () => $("ops-detail").close());
  $("ops-detail").addEventListener("close", () => {detailGeneration++; $("ops-detail-content").replaceChildren();});
  window.addEventListener("pagehide", () => {generation++; detailGeneration++; $("ops-detail-content").replaceChildren();});
  refresh();
})();

File: src/cloud_soc/portal/operations.py
"""Read-only operational totals and exact evidence, never synthetic incidents."""

from datetime import datetime, timezone
import re
from elasticsearch import NotFoundError
from cloud_soc.portal.agent_status import iso, parse_time
from cloud_soc.portal.log_contract import INDICES, SOURCE_FIELDS, field, parse_filters, document_reference, project_hit
from cloud_soc.portal.log_query import LogQueryError, bounded_json
from cloud_soc.portal.security_detail import DETAIL_FIELDS, security_detail, text
from cloud_soc.processing.contract import NORMALIZED, RECORDS, STATUS
from cloud_soc.processing.worker import checked

ALERT = "security-alerts"
ALERT_FIELDS = ["@timestamp", "rule.id", "rule.name", "organization.id", "source.ip", "cloud_soc.severity",
                "cloud_soc.alert_title", "cloud_soc.event_count", "cloud_soc.threshold", "cloud_soc.time_window_seconds",
                "cloud_soc.window_start", "cloud_soc.window_end", "cloud_soc.provenance"]


def exact_indices(client, patterns):
    try:
        response = client.indices.resolve_index(name=",".join(patterns), expand_wildcards="open")
    except NotFoundError:
        return []
    if response.get("aliases") or response.get("data_streams"):
        raise ValueError("Aliases are not accepted")
    names = [row["name"] for row in response.get("indices", [])]
    for name in names:
        if not any(re.fullmatch(re.escape(pattern).replace(r"\*", ".*"), name) for pattern in patterns):
            raise ValueError("Unexpected index")
    return names


def identifier(value):
    if not isinstance(value, str) or not value or len(value) > 512 or text(value) != value:
        raise ValueError("Invalid document ID")
    return value


def alert_row(hit):
    source = hit["_source"]
    get = lambda path: text(field(source, path))
    return {"id": identifier(hit["_id"]), "timestamp": get("@timestamp"), "rule": get("rule.id"),
            "title": get("cloud_soc.alert_title") or get("rule.name"), "severity": get("cloud_soc.severity"),
            "organization": get("organization.id"), "source_ip": get("source.ip")}


class Operations:
    def __init__(self, client):
        self.client = client.options(request_timeout=5, max_retries=0) if client is not None else None

    def get(self, index, doc_id, fields):
        if not self.client:
            raise RuntimeError("Monitor not configured")
        if not exact_indices(self.client, [index]):
            return None
        try:
            return self.client.get(index=index, id=identifier(doc_id), source_includes=fields)["_source"]
        except NotFoundError:
            return None

    def series(self, patterns, time_field, start, end, interval, *, records=False, alerts=False):
        if not self.client:
            raise RuntimeError()
        names = exact_indices(self.client, patterns)
        if not names:
            return {"state": "no_index", "count": 0, "buckets": [], "rows": [], "statuses": []}
        caps = self.client.field_caps(index=names, fields=[time_field], include_unmapped=True)["fields"].get(time_field, {})
        if not caps or set(caps) - {"date", "date_nanos"}:
            raise ValueError("Time mapping missing")
        aggs = {"trend": {"date_histogram": {"field": time_field, "fixed_interval": interval,
                    "min_doc_count": 0, "extended_bounds": {"min": start, "max": iso(parse_time(end))}}}}
        if records:
            aggs["status"] = {"terms": {"field": "status", "size": 10}}
        result = checked(self.client.search(index=names, size=50 if alerts else 0,
            source=ALERT_FIELDS if alerts else False, sort=[{"@timestamp": "desc"}] if alerts else None,
            track_total_hits=True, timeout="4s", allow_partial_search_results=False,
            query={"range": {time_field: {"gte": start, "lt": end}}}, aggs=aggs))
        if result["hits"]["total"]["relation"] != "eq":
            raise ValueError("Inexact total")
        return {"state": "ok", "count": result["hits"]["total"]["value"],
                "buckets": [{"time": b["key_as_string"], "count": b["doc_count"]} for b in result["aggregations"]["trend"]["buckets"]],
                "rows": [alert_row(hit) for hit in result["hits"]["hits"]] if alerts else [],
                "statuses": [{"key": b["key"] if b["key"] in ("normalized", "unsupported", "invalid") else "unknown",
                              "doc_count": b["doc_count"]} for b in result["aggregations"].get("status", {}).get("buckets", [])]}

    def pipeline(self):
        value = self.get(STATUS, "normalizer", ["@timestamp", "state", "last_success", "checkpoint", "error", "detection"])
        if value is None:
            return {"state": "not_started", "detection": "disabled_pending_approval"}
        stamp = parse_time(value.get("@timestamp"))
        age = (datetime.now(timezone.utc) - stamp).total_seconds() if stamp else None
        state = value.get("state")
        return {"state": state if age is not None and 0 <= age <= 300 and state in ("success", "running", "failed", "waiting") else "stale",
                "last_success": text(value.get("last_success")), "checkpoint": text(value.get("checkpoint")),
                "detection": "disabled_pending_approval"}

    def summary(self, pairs):
        pairs = list(pairs)
        try:
            if any(key not in ("start", "end") for key, _ in pairs):
                raise ValueError()
            filters = parse_filters(pairs)
        except ValueError:
            raise LogQueryError("invalid_operations_query", 400, "조회 기간을 확인하세요.") from None
        interval = "5m" if (parse_time(filters.end) - parse_time(filters.start)).total_seconds() <= 86400 else "1h"
        result = {"start": filters.start, "end": filters.end, "incident_work": "not_implemented", "sections": {}}
        for key, operation in {
            "intake": lambda: self.series(INDICES, "event.ingested", filters.start, filters.end, interval),
            "processing": lambda: self.series([RECORDS], "event.ingested", filters.start, filters.end, interval, records=True),
            "alerts": lambda: self.series([ALERT], "@timestamp", filters.start, filters.end, interval, alerts=True),
            "quality": lambda: self.series(["soc-agent-health-*"], "event.ingested", filters.start, filters.end, interval),
            "pipeline": self.pipeline,
        }.items():
            try:
                result["sections"][key] = operation()
            except Exception:
                result["sections"][key] = {"state": "unavailable"}
        return bounded_json(result)

    def detail(self, pairs):
        try:
            pairs = list(pairs)
            values = dict(pairs)
            if len(pairs) != len(values) or set(values) not in ({"id"}, {"id", "evidence"}):
                raise ValueError()
            doc_id = identifier(values["id"])
            position = values.get("evidence")
            if position is not None and not re.fullmatch(r"[0-9]{1,3}", position):
                raise ValueError()
        except ValueError:
            raise LogQueryError("invalid_alert_query", 400, "경보 참조를 확인하세요.") from None
        try:
            source = self.get(ALERT, doc_id, ALERT_FIELDS)
            if source is None:
                raise LogQueryError("alert_missing", 404, "경보가 없거나 보존 기간이 지났습니다.")
            evidence = field(source, "cloud_soc.provenance.evidence")
            evidence = evidence if isinstance(evidence, list) else []
            result = {"alert": alert_row({"_source": source, "_id": doc_id}), "evidence_count": len(evidence),
                      "evidence_limit": min(len(evidence), 100), "evidence_state": "referenced" if evidence else "legacy_missing_provenance",
                      "rule_version": text(field(source, "cloud_soc.provenance.rule_version")),
                      "engine_version": text(field(source, "cloud_soc.provenance.engine_version")),
                      "condition": {k: field(source, "cloud_soc." + k) for k in ("event_count", "threshold", "time_window_seconds")
                                    if type(field(source, "cloud_soc." + k)) is int}}
            if position is not None:
                number = int(position)
                if number >= min(len(evidence), 100):
                    raise LogQueryError("evidence_missing", 404, "요청한 근거 참조가 없습니다.")
                result["evidence"] = self.evidence(evidence[number])
            return bounded_json(result)
        except LogQueryError:
            raise
        except Exception:
            raise LogQueryError("alerts_unavailable", 503, "경보·근거를 조회하지 못했습니다. 권한과 참조를 확인하세요.") from None

    def evidence(self, reference):
        if not isinstance(reference, dict):
            return {"state": "invalid_reference"}
        normalized, raw = reference.get("normalized"), reference.get("raw")
        try:
            if not isinstance(normalized, dict) or normalized.get("index") not in (NORMALIZED, "normalized-events"):
                raise ValueError()
            normalized = {"index": normalized["index"], "id": identifier(normalized.get("id"))}
            if isinstance(raw, dict) and re.fullmatch(r"raw-logs-[a-z0-9][a-z0-9_.-]{0,220}", str(raw.get("index", ""))):
                raw = {"index": raw["index"], "id": identifier(raw.get("id"))}
            else:
                raw = document_reference(raw.get("index"), raw.get("id"))
        except (ValueError, AttributeError):
            return {"state": "unsupported_or_invalid_reference"}
        norm = self.get(normalized["index"], normalized["id"], ["@timestamp", "event.category", "event.action", "event.outcome", "cloud_soc.provenance.raw"])
        if norm is None:
            return {"state": "normalized_missing_or_expired", "normalized": normalized, "raw": raw}
        if field(norm, "cloud_soc.provenance.raw") != raw:
            return {"state": "provenance_mismatch"}
        source = self.get(raw["index"], raw["id"], list(SOURCE_FIELDS + DETAIL_FIELDS))
        if source is None:
            return {"state": "raw_missing_or_expired", "normalized": normalized, "raw": raw}
        metadata = ({path: text(field(source, path)) for path in ("@timestamp", "host.name", "user.name", "event.action")}
                    if raw["index"].startswith("raw-logs-") else project_hit({"_index": raw["index"], "_id": raw["id"], "_source": source}))
        return {"state": "exact_reference", "normalized": normalized, "raw": raw,
                "metadata": metadata,
                "security": security_detail(source)}
